refactor: deprecate AuthToken for non-OG pyroscope cloud

marcsanmi opened the pull request:
Closes https://github.com/grafana/pyroscope-go/issues/124

This PR deprecates the use of `AuthToken` for non-OG Pyroscope cloud.

Changes:
- Added deprecation notices to `AuthToken` in Config structs
- Updated authentication logic in `uploadProfile` function
- Added warning for deprecated `AuthToken` usage with non-OG Pyroscope cloud
- Maintained `AuthToken` support for OG Pyroscope cloud servers
- Ensured BasicAuth is used when provided, regardless of server type

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/grafana/pyroscope-go?pullRequest=125) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/grafana/pyroscope-go?pullRequest=125) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/grafana/pyroscope-go?pullRequest=125) it.</sub>

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/grafana/pyroscope-go?pullRequest=125) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/grafana/pyroscope-go?pullRequest=125) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/grafana/pyroscope-go?pullRequest=125) it.</sub>

korniltsev commented on upstream/remote/remote.go:
```diff
@@ -188,10 +190,12 @@ func (r *Remote) uploadProfile(j *upstream.UploadJob) error {
 	} else if r.cfg.BasicAuthUser != "" && r.cfg.BasicAuthPassword != "" {
 		request.SetBasicAuth(r.cfg.BasicAuthUser, r.cfg.BasicAuthPassword)
+	} else if r.cfg.AuthToken != "" {
+		r.logger.Infof("Deprecated: AuthToken is being used. Please switch to BasicAuthUser and BasicAuthPassword.")
```
>  AuthToken is being used
It says the token is beeing used, but in fact it is ignored. Maybe say something like "Authtoken is specified, but deprecated and ignored Please switch to ..."

Also if someone still wants to use Bearer (for example they put private pyroscope behind nginx with auth), they can use `HTTPHeaders `, maybe it is also worth mentioning in the warning.

marcsanmi replied:
Btw, I set it as `Info` not `Warn` since `Logger` [interface](https://github.com/grafana/pyroscope-go/blob/main/upstream/remote/remote.go#L55) doesn't implement it but I could if needed, lmkwyt.

korniltsev replied:
I think it is fine.  I think if we add a a method to logger  interface it will break compilation for all users who has custom logger implementation

korniltsev commented:
nit: maybe add a test if possible and you think it may be useful, but I don't think it is required

marcsanmi commented:
I added tests for the `uploadProfile` function in the remote package, covering various authentication scenarios. I also Introduced a new `internal/testutil` package with a TestLogger for shared testing utilities, following a similar approach used in projects like Prometheus.

korniltsev commented:
Awesome

Fabio1988 commented on upstream/remote/remote.go:
```diff
@@ -188,10 +196,12 @@ func (r *Remote) uploadProfile(j *upstream.UploadJob) error {
 	} else if r.cfg.BasicAuthUser != "" && r.cfg.BasicAuthPassword != "" {
 		request.SetBasicAuth(r.cfg.BasicAuthUser, r.cfg.BasicAuthPassword)
+	} else if r.cfg.AuthToken != "" {
+		r.logger.Infof(authTokenDeprecationWarning)
```
that's a bold move. you deprecate the usage of 'AuthToken' but in the same commit you simply removed the usage for self-hosting environments :( 
this broke our setup, downgrade was required afterwards :D
